src/stroke.rs

```diff
@@ -0,0 +1,254 @@
+    pub fn with_start_cap(mut self, cap: Cap) -> Self {
+        self.start_cap = cap;
+        self
+    }
+
+    /// Builder method for setting the cap style for the end of the stroke.
+    pub fn with_end_cap(mut self, cap: Cap) -> Self {
+        self.end_cap = cap;
+        self
+    }
+
+    /// Builder method for setting the cap style.
+    pub fn with_caps(mut self, cap: Cap) -> Self {
+        self.start_cap = cap;
+        self.end_cap = cap;
+        self
+    }
+
+    /// Builder method for setting the dashing parameters.
+    pub fn with_dashes<P>(mut self, offset: f64, pattern: P) -> Self
+    where
+        P: IntoIterator,
+        P::Item: Borrow<f64>,
+    {
+        self.dash_offset = offset;
+        self.dash_pattern.clear();
+        self.dash_pattern
+            .extend(pattern.into_iter().map(|dash| *dash.borrow()));
+        self
+    }
+
+    /// Builder method for setting whether or not the stroke should be affected
+    /// by the scale of any applied transform.
+    pub fn with_scale(mut self, yes: bool) -> Self {
+        self.scale = yes;
+        self
+    }
+}
+
+/// Collection of values representing lengths in a dash pattern.
+pub type Dashes = SmallVec<[f64; 4]>;
+
+/// Internal structure used for creating strokes.
+#[derive(Default)]
+struct StrokeCtx {
+    // Probably don't need both output and forward, can just concat
+    output: BezPath,
+    forward_path: BezPath,
+    backward_path: BezPath,
+    last_pt: Point,
+    last_tan: Vec2,
+}
+
+/// Expand a stroke into a fill.
+pub fn stroke(path: impl IntoIterator<Item = PathEl>, style: &Stroke, tolerance: f64) -> BezPath {
+    let mut ctx = StrokeCtx::default();
+    for el in path {
+        let p0 = ctx.last_pt;
+        match el {
+            PathEl::MoveTo(p) => {
+                ctx.finish();
+                ctx.last_pt = p;
+            }
+            PathEl::LineTo(p1) => {
+                if p1 != ctx.last_pt {
+                    let tangent = p1 - p0;
+                    ctx.do_tangents(style, tangent, tangent, p1);
+                    ctx.do_line(style, tangent, p1);
+                }
+            }
+            PathEl::QuadTo(p1, p2) => {
+                if p1 != p0 && p2 != p0 {
+                    let q = QuadBez::new(p0, p1, p2);
+                    let (tan0, tan1) = PathSeg::Quad(q).tangents();
+                    ctx.do_tangents(style, tan0, tan1, p2);
+                    ctx.do_cubic(style, q.raise(), tolerance);
+                }
+            }
+            PathEl::CurveTo(p1, p2, p3) => {
+                if p1 != p0 && p2 != p0 && p3 != p0 {
+                    let c = CubicBez::new(p0, p1, p2, p3);
+                    let (tan0, tan1) = PathSeg::Cubic(c).tangents();
+                    ctx.do_tangents(style, tan0, tan1, p3);
+                    ctx.do_cubic(style, c, tolerance);
+                }
+            }
+            _ => todo!(),
+        }
+    }
+    todo!()
+}
+
+fn get_end(el: &PathEl) -> Point {
```

Comment:
I think there should be an `end_point` method on `PathEl` that returns `Option<Point>` and is implemented exactly like this.